src/libjson/json.c

```diff
@@ -0,0 +1,1217 @@
+
+            CHECK(_invoke_callback(parser, JSON_REASON_VALUE, type, &un));
+            break;
+        }
+    }
+
+done:
+    return result;
+}
+
+json_result_t json_parser_init(
+    json_parser_t* parser,
+    char* data,
+    size_t size,
+    json_parser_callback_t callback,
+    void* callback_data,
+    json_allocator_t* allocator)
+{
+    if (!parser || !data || !size || !callback)
+        return JSON_BAD_PARAMETER;
+
+    if (!allocator || !allocator->ja_malloc || !allocator->ja_free)
+        return JSON_BAD_PARAMETER;
+
+    _memset(parser, 0, sizeof(json_parser_t));
+    parser->data = data;
+    parser->ptr = data;
+    parser->end = data + size;
+    parser->callback = callback;
+    parser->callback_data = callback_data;
+
+    return JSON_OK;
+}
+
+json_result_t json_parser_parse(json_parser_t* parser)
+{
+    json_result_t result = JSON_OK;
+    char c;
+
+    /* Check parameters */
+    if (!parser)
+        return JSON_BAD_PARAMETER;
+
+    /* Expect '{' */
+    {
+        /* Skip whitespace */
+        while (parser->ptr != parser->end && _isspace(*parser->ptr))
+            parser->ptr++;
+
+        /* Fail if output exhausted */
+        if (parser->ptr == parser->end)
+            RAISE(JSON_EOF);
+
+        /* Read the next character */
+        c = *parser->ptr++;
+
+        /* Expect object-begin */
+        if (c != '{')
+            return JSON_BAD_SYNTAX;
+    }
+
+    CHECK(_get_object(parser));
+
+done:
+    return result;
+}
+
+static int _strtou64(uint64_t* x, const char* str)
+{
+    char* end;
+
+    *x = _strtoul(str, &end, 10);
+
+    if (!end || *end != '\0')
+        return -1;
+
+    return 0;
+}
+
+json_result_t json_match(
+    json_parser_t* parser,
+    const char* pattern,
+    unsigned long* index)
+{
+    json_result_t result = JSON_UNEXPECTED;
+    char buf[256];
+    char* ptr = NULL;
+    const char* pattern_path[JSON_MAX_NESTING];
+    size_t pattern_depth = 0;
+    unsigned long n = 0;
+    size_t pattern_len;
+
+    if (!parser || !parser->path || !pattern)
+        RAISE(JSON_BAD_PARAMETER);
+
+    /* Make a copy of the pattern that can be modified */
+    {
+        pattern_len = _strlen(pattern);
+
+        if (pattern_len < sizeof(buf))
+            ptr = buf;
+        else if (!(ptr = parser->allocator->ja_malloc(pattern_len + 1)))
+            RAISE(JSON_OUT_OF_MEMORY);
+
+        _strlcpy(ptr, pattern, pattern_len + 1);
+    }
+
+    /* Split the pattern into tokens */
+    if ((pattern_depth = _split(ptr, '.', pattern_path,
+        JSON_MAX_NESTING)) == (size_t)-1)
+    {
+        RAISE(JSON_NESTING_OVERFLOW);
+    }
+
+    /* Return false if the path sizes are different */
+    if (parser->depth != pattern_depth)
+    {
+        result = JSON_NO_MATCH;
+        goto done;
+    }
+
+    /* Compare the elements */
+    for (size_t i = 0; i < pattern_depth; i++)
+    {
+        if (_strcmp(pattern_path[i], "#") == 0)
+        {
+            if (_strtou64(&n, parser->path[i]) != 0)
```

Comment:
It would be nice if we could get the index/`n` for arrays too and I think it wouldn't be too hard to add. I think we'd just need an additional `size_t` for every path element (or a separate vector with the same nesting as the path).

Reply:
Good idea. Done. The index is now included in the **parser->nodes[]** array.